Improve parse_declarations! performance

leonid-shevtsov opened the pull request:
This PR replaces regexp-matching code in `parse_declarations!` with equivalent substring lookup using `.split(&block)` and `.index`.

On our benchmarks this shows a 10% improvement in performance measured for _overall_ `add_rule!` / `add_block!`. So, quite a significant one.

## Pre-Merge Checklist

- [x] CHANGELOG.md updated with short summary

grosser commented on lib/css_parser/rule_set.rb:
```diff
@@ -647,14 +652,21 @@ def parse_declarations!(block) # :nodoc:
-          value = matches[2]
+      block.split(SEMICOLON) do |decs|
+        decs = (continuation ? "#{continuation};#{decs}" : decs)
+        lparen_index = decs.index(LPAREN)
+        if !lparen_index.nil? && decs.index(RPAREN, lparen_index).nil? # if it has an unmatched parenthesis
```
can these 2 lines go into a descriptive method ?

grosser commented on lib/css_parser/rule_set.rb:
```diff
@@ -647,14 +652,21 @@ def parse_declarations!(block) # :nodoc:
       return unless block
 
       continuation = nil
-      block.split(/[;$]+/m).each do |decs|
```
this was splitting on newline or semicolon, is this still there ?
did you find good test coverage or should we add some more for edge-cases before this refactor, for example "\n\n;;;\n\n" and other nonsense to make sure it does not blow up

leonid-shevtsov commented:
@grosser I've addressed the comments.

As to the splitting by newline - as I recall, I removed it because newlines are not significant in CSS; only semicolons can split declarations. So, they can be handled as any other whitespace (stripped, that is.)

I added a test, if you have something else in mind, let's test that as well.

grosser commented on lib/css_parser/rule_set.rb:
```diff
@@ -647,20 +652,31 @@ def parse_declarations!(block) # :nodoc:
       end
     end
 
+    def unmatched_parenthesis?(declarations)
```
reads very nicely!

maybe more descriptive:
```suggestion
    def unmatched_open_parenthesis?(declarations)
```

should avoid `.nil?` for [performance reasons](https://stackoverflow.com/questions/1972266/obj-nil-vs-obj-nil)

```
(lparen_index = declarations.index(LPAREN)) && !declarations.index(RPAREN, lparen_index)
```

grosser commented on lib/css_parser/rule_set.rb:
```diff
@@ -647,20 +652,31 @@ def parse_declarations!(block) # :nodoc:
+        if unmatched_parenthesis?(decs)
+          continuation = decs
+        else
+          colon = decs.index(COLON)
+          next if colon.nil?
```
```suggestion
          next unless (colon = decs.index(COLON))
```

grosser commented on lib/css_parser/rule_set.rb:
```diff
@@ -647,20 +652,31 @@ def parse_declarations!(block) # :nodoc:
+          next if colon.nil?
+
+          property = decs[0, colon]
+          value = decs[colon + 1, decs.length - colon - 1]
```
is that what you wanted ?
```suggestion
          value = decs[(colon + 1)..]
```

leonid-shevtsov commented:
Done @grosser.

grosser commented:
1.19.0
